# Module 2: Practical Tools for Machine Learning

## Lecture 1: *Handling Data with PANDAS*

# Anaconda - Open-source Python/R Distribution for Scientific Computing

For installation of Anaconda, Refer the tutorial at: 
https://www.datacamp.com/community/tutorials/installing-anaconda-windows

#### 1. Scientific Computing: Data Science, Machine Learning, Deep Learning, Statistical Analysis
#### 2. Includes Python & R,  many packages such as pandas, numpy, scikit-learn, tensorflow

#### 3. Platform agnostic - Windows, MacOS, Linux


## Jupiter notebook is a web app for creating and sharing computational documents. 
#### Launch Jupiter notebook from Anaconda
#### Has live code, textual descriptions, equations, visualization



## Introduction to Pandas Library

pandas - a powerful data analysis and manipulation library for Python
=============================================================

    - Create, Load, merge data
    
    - Clean data: complete missing values, remove duplicates, reformat 
    
    - Manipulate Data: perform data analytics, provide time-series functions
    
    - Data Analysis: correlate and Visualize data


In [1]:
import pandas as pd

## 1) Pandas Series

#### Creating Series :-
    
    1) From Scalar Value 
    
    2) From Dictionary

## 1.1) Scalar Series 

*** Scaler series must specify both data and index***

In [2]:
s1 = pd.Series(data = [5.0,3,4,7,8])
s1

0    5.0
1    3.0
2    4.0
3    7.0
4    8.0
dtype: float64

In [3]:
s2 = pd.Series(data = [2,3,4,5,6],index = [0,1,2,3,4])
s2

0    2
1    3
2    4
3    5
4    6
dtype: int64

In [4]:
type(s2.index)

pandas.core.indexes.numeric.Int64Index

*** If single data is specified, it will be repeated to match index length***

In [5]:
s3 = pd.Series(5.9,[1,2,3,4,5])
s3

1    5.9
2    5.9
3    5.9
4    5.9
5    5.9
dtype: float64

#### A series is like a vector. Vectorized operations with Series:

In [6]:
s2+s3

0     NaN
1     8.9
2     9.9
3    10.9
4    11.9
5     NaN
dtype: float64

In [7]:
s2*s3

0     NaN
1    17.7
2    23.6
3    29.5
4    35.4
5     NaN
dtype: float64

In [8]:
s2**2

0     4
1     9
2    16
3    25
4    36
dtype: int64

In [9]:
s2/s3

0         NaN
1    0.508475
2    0.677966
3    0.847458
4    1.016949
5         NaN
dtype: float64

In [10]:
s2-s3

0    NaN
1   -2.9
2   -1.9
3   -0.9
4    0.1
5    NaN
dtype: float64

## 1.2. Series from a Dictionary

**When the data is a dictionary {key: value}, values become members of the series, keys become their index.**



In [11]:
d = {"a":1,"b":2,"c":3,"d":4}
s4 = pd.Series(d)
s4

a    1
b    2
c    3
d    4
dtype: int64

**You can confirm the data type of the series**

In [12]:
type(s4)

pandas.core.series.Series

**If an index is passed as list-of-labels, the values in dictionary corresponding to the labels index will be pulled out.**

In [13]:
s5 = pd.Series(d,index = ['b',"c","a","d"])
s5

b    2
c    3
a    1
d    4
dtype: int64

**An extra index will generate Not A Number NaN**

In [14]:
s6 = pd.Series(d,index = ['b',"c","a","d","e","f"])
s6

b    2.0
c    3.0
a    1.0
d    4.0
e    NaN
f    NaN
dtype: float64

**If only 1 data item is specified, it will be replicated for all index positions**

***You can get and set values by referring an index label in any series***

In [15]:
s6["b"]

2.0

In [16]:
s6["e"] = 5.0
s6

b    2.0
c    3.0
a    1.0
d    4.0
e    5.0
f    NaN
dtype: float64

## 2. Pandas DataFrames

### 2.1) Creating DataFrames

DataFrame is a **2-dimensional** labeled data structure with **columns that can be of different types**. It just like a spreadsheet or SQL table, or a dict of Series objects. 

***Method of creating a DataFrame***

d = { 
        "column_name" : pd.Series(data,index = ["row_name"])
    }

df = pd.DataFrame(d)

#### a) Creating a Dataframe from dictionary of series 

In [17]:
ser1 =pd.Series([9,10],[1,2])
ser2 =pd.Series([9,12],[4,5])
dict1={3:ser1,4:ser1}
df1 = pd.DataFrame(dict1)
df1

3   4
1   9   9
2  10  10

In [18]:
dict2={5:ser2,6:ser1}
df2 = pd.DataFrame(dict2)
df2

5     6
1   NaN   9.0
2   NaN  10.0
4   9.0   NaN
5  12.0   NaN

In [19]:
type(df2)

pandas.core.frame.DataFrame

#### For enumerating columns of Dataframe

In [20]:
df2.columns

Int64Index([5, 6], dtype='int64')

#### For enumerating row indices of Dataframe

In [21]:
df2.index

Int64Index([1, 2, 4, 5], dtype='int64')

##### For enumerating both axes - (columns and rows) - DataFrame.axes



In [22]:
df2.axes

[Int64Index([1, 2, 4, 5], dtype='int64'), Int64Index([5, 6], dtype='int64')]

#### For accessing a column

In [23]:
df2[5]

1     NaN
2     NaN
4     9.0
5    12.0
Name: 5, dtype: float64

#### For accessing an element

In [24]:
df2[5][4]

9.0

#### b) Creating a Dataframe from a list of dictionaries

In [25]:
data = [{'a' : 10,'b' : 20 , 'c' : 25}, {'a' : 11 , 'b' : 18}]

In [26]:
df = pd.DataFrame(data)
df

a   b     c
0  10  20  25.0
1  11  18   NaN

####  c) Creating a Dataframe from an indexed list of dictionaries. Note shape must match original number of dictionaries

In [27]:
df = pd.DataFrame(data, index = ['first','second'])
df

a   b     c
first   10  20  25.0
second  11  18   NaN

### 2.2) DataFrame Basic functionalities 

In [28]:
df

a   b     c
first   10  20  25.0
second  11  18   NaN

##### a) Transpose - DataFrame.T

It returns the transpose of the dataframe.

In [29]:
df.T

first  second
a   10.0    11.0
b   20.0    18.0
c   25.0     NaN

#### b) DataFrame.dtypes
It return the data types of each columns.

In [30]:
df.dtypes

a      int64
b      int64
c    float64
dtype: object

##### c) DataFrame.empty
It return the boolean value saying whether the object is empty or not.**True** indicates that the object is empty

In [31]:
df.empty

False

##### d) DataFrame.shape
It return a tuple representing the dimensional of dataframe.It return in the form of tuple(no._of_rows,no_of_columns)

In [32]:
df.shape

(2, 3)

##### e) DataFrame.size
It return the number of elements in the dataframe.

In [33]:
df.size

6

#### f) DataFrame.values
It returns the actual data in the dataframe as array format.

In [34]:
df.values

array([[10., 20., 25.],
       [11., 18., nan]])

**g) DataFrame.isnull()**

It return which column contain NaN value in term of true and false

In [35]:
df.isnull()

a      b      c
first   False  False  False
second  False  False   True

In [36]:
df.isnull().sum() #count the NaN value

a    0
b    0
c    1
dtype: int64

## 3. Pandas :- file read and file write operations

#### a) Read CSV file

In [37]:
df = pd.read_csv("workingfile.csv")
df

ID first_name company  salary
0  11      David     Aon      74
1  12      Jamie     TCS      76
2  13      Steve  Google      96
3  14    Stevart     RBS      71
4  15       John       .      78

#### b) read HTML file

In [38]:
url = "http://www.basketball-reference.com/leagues/NBA_2015_totals.html"
df = pd.read_html(url)
df

[      Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
 0      1      Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79   
 1      2    Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   
 2      3    Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199   
 3      4     Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   
 4      5   Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27   
 ..   ...             ...  ..  ..  ...  ..  ..   ...  ...  ...  ...   ...  ...   
 670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
 671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
 672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
 673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
 674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   
 
      DRB  TRB

**HTML loads in multiple frames. we can assces any frame**

In [39]:
df = df[0]   # It is the first dataframe in the list of dataframes returned by read_html()


#### If u want to read bottom 5 rows

In [40]:
df.tail()

Rk          Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  \
670  490  Thaddeus Young  PF  26  TOT  76  68  2434  451  968  ...  .655  127   
671  490  Thaddeus Young  PF  26  MIN  48  48  1605  289  641  ...  .682   75   
672  490  Thaddeus Young  PF  26  BRK  28  20   829  162  327  ...  .606   52   
673  491     Cody Zeller   C  22  CHO  62  45  1487  172  373  ...  .774   97   
674  492    Tyler Zeller   C  25  BOS  82  59  1731  340  619  ...  .823  146   

     DRB  TRB  AST  STL BLK  TOV   PF   PTS  
670  284  411  173  124  25  117  171  1071  
671  170  245  135   86  17   75  115   685  
672  114  166   38   38   8   42   56   386  
673  265  362  100   34  49   62  156   472  
674  319  465  113   18  52   76  205   833  

[5 rows x 30 columns]

#### If u want to read top 5 rows

In [41]:
df.head()

Rk         Player Pos Age   Tm   G  GS    MP   FG  FGA  ...   FT%  ORB  DRB  \
0  1     Quincy Acy  PF  24  NYK  68  22  1287  152  331  ...  .784   79  222   
1  2   Jordan Adams  SG  20  MEM  30   0   248   35   86  ...  .609    9   19   
2  3   Steven Adams   C  21  OKC  70  67  1771  217  399  ...  .502  199  324   
3  4    Jeff Adrien  PF  28  MIN  17   0   215   19   44  ...  .579   23   54   
4  5  Arron Afflalo  SG  29  TOT  78  72  2502  375  884  ...  .843   27  220   

   TRB  AST STL BLK  TOV   PF   PTS  
0  301   68  27  22   60  147   398  
1   28   16  16   7   14   24    94  
2  523   66  38  86   99  222   537  
3   77   15   4   9    9   30    60  
4  247  129  41   7  116  167  1035  

[5 rows x 30 columns]

#### c) read tsv file

In [42]:
df = pd.read_table("test.tsv")
df

test     test.1     test.2     test.3     test.4
0      vikas      vikas      vikas      vikas      vikas
1  Parjapati  Parjapati  Parjapati  Parjapati  Parjapati
2     Aditya     Aditya     Aditya     Aditya     Aditya
3      kumar      kumar      kumar      kumar      kumar

#### d) read json file

In [43]:
df = pd.read_json("example.json")
df

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

#### e) read excel file

In [44]:
df = pd.read_excel("ex1.xlsx")
#df = pd.read_excel("NFHS_5_India_Districts_Factsheet_Data.xls")
df

Unnamed: 0  a   b   c   d message
0           0  1   2   3   4   hello
1           1  5   6   7   8   world
2           2  9  10  11  12     foo

#### f) read data from  github repository 

In [45]:
data = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## 4. Initial Data Analysis

## Unemployment Rate by Social Status (Rural) as a percentage of Labour Force

In [46]:
## Import dataset 
data = pd.read_csv("data_gov.csv")
data.head() # Show top five row

Category of States Non Special Category States  \
0  Non Special Category States              Andhra Pradesh   
1  Non Special Category States                       Assam   
2  Non Special Category States                       Bihar   
3  Non Special Category States                Chhattisgarh   
4  Non Special Category States                       Delhi   

   Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                        17.1                        12.8   
1                         9.4                         9.5   
2                        10.6                        10.0   
3                        12.5                         4.3   
4                         NaN                         NaN   

   Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
0                        12.0                         8.2   
1                         4.9                         7.7   
2                        11.2                         7.7   
3                         4.2                         5.8   
4                         NaN                         NaN   

   Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
0                              10.1                               9.7   
1                               4.3                               6.9   
2                               5.2                               7.0   
3                              10.2                               8.2   
4                               1.5                               4.8   

   Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
0               7.4               5.6                         11.2   
1               7.3               6.4                          6.4   
2               5.9               5.3                          6.8   
3               8.1               9.0                          8.2   
4               2.6               2.9                          1.9   

   All Social Groups - 2007-08  
0                          9.5  
1                          7.0  
2                          7.4  
3                          6.7  
4                          3.8

In [47]:
# Show bottom five rows
data.tail()

Category of States Non Special Category States  Scheduled Castes - 2004-05  \
31  Union Territories        Dadra & Nagar Haveli                         NaN   
32  Union Territories                 Daman & Diu                         NaN   
33  Union Territories                 Lakshadweep                         NaN   
34  Union Territories                  Puducherry                        36.0   
35          All India                   All India                        12.0   

    Scheduled Castes - 2007-08  Scheduled Tribes - 2004-05  \
31                         NaN                         7.1   
32                         NaN                         NaN   
33                         NaN                        25.1   
34                        28.7                         NaN   
35                        11.9                         6.5   

    Scheduled Tribes - 2007-08  Other Backward Classes - 2004-05  \
31                         8.4                               NaN   
32                         NaN                               NaN   
33                        17.0                               NaN   
34                         NaN                              22.2   
35                         7.5                               7.7   

    Other Backward Classes - 2007-08  Others - 2004-05  Others - 2007-08  \
31                               NaN               NaN               NaN   
32                               NaN               NaN               NaN   
33                               NaN               NaN               NaN   
34                              22.8               NaN               NaN   
35                               7.9               6.6               6.4   

    All Social Groups - 2004-05  All Social Groups - 2007-08  
31                          6.0                          6.2  
32                          1.6                          1.7  
33                         22.5                         16.7  
34                         27.4                         25.2  
35                          8.2                          8.4

In [48]:
## shape of dataset
data.shape

(36, 12)

there are 36 no. of rows and 12 no. of columns in our dataset.

In [49]:
## size of data set
data.size

432

there are 432 no. of items in our dataset.

In [50]:
### Checking the data type of columns
data.dtypes

Category of States                   object
Non Special Category States          object
Scheduled Castes - 2004-05          float64
Scheduled Castes - 2007-08          float64
Scheduled Tribes - 2004-05          float64
Scheduled Tribes - 2007-08          float64
Other Backward Classes - 2004-05    float64
Other Backward Classes - 2007-08    float64
Others - 2004-05                    float64
Others - 2007-08                    float64
All Social Groups - 2004-05         float64
All Social Groups - 2007-08         float64
dtype: object

In [51]:
#### Fetch Information from dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Category of States                36 non-null     object 
 1   Non Special Category States       36 non-null     object 
 2   Scheduled Castes - 2004-05        26 non-null     float64
 3   Scheduled Castes - 2007-08        26 non-null     float64
 4   Scheduled Tribes - 2004-05        27 non-null     float64
 5   Scheduled Tribes - 2007-08        27 non-null     float64
 6   Other Backward Classes - 2004-05  27 non-null     float64
 7   Other Backward Classes - 2007-08  27 non-null     float64
 8   Others - 2004-05                  29 non-null     float64
 9   Others - 2007-08                  29 non-null     float64
 10  All Social Groups - 2004-05       36 non-null     float64
 11  All Social Groups - 2007-08       36 non-null     float64
dtypes: float64

In [52]:
## Fetching data for a specific column
data["Category of States"]

0     Non Special Category States
1     Non Special Category States
2     Non Special Category States
3     Non Special Category States
4     Non Special Category States
5     Non Special Category States
6     Non Special Category States
7     Non Special Category States
8     Non Special Category States
9     Non Special Category States
10    Non Special Category States
11    Non Special Category States
12    Non Special Category States
13    Non Special Category States
14    Non Special Category States
15    Non Special Category States
16    Non Special Category States
17    Non Special Category States
18    Non Special Category States
19        Special Category States
20        Special Category States
21        Special Category States
22        Special Category States
23        Special Category States
24        Special Category States
25        Special Category States
26        Special Category States
27        Special Category States
28        Special Category States
29            

In [53]:
### collecting unique values (non duplicated) for a specific olumn
data["Category of States"].unique()

array(['Non Special Category States', 'Special Category States',
       'Union Territories', 'All India'], dtype=object)

In [54]:
### seeing number of unique values (non duplicated) for a specific olumn
len(data["Category of States"].unique())

4

In [55]:
### fetching data values of specific columns from the dataset
data[['Scheduled Castes - 2004-05', 'Scheduled Castes - 2007-08','Scheduled Tribes - 2004-05', 'Scheduled Tribes - 2007-08']]

Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
0                         17.1                        12.8   
1                          9.4                         9.5   
2                         10.6                        10.0   
3                         12.5                         4.3   
4                          NaN                         NaN   
5                         19.2                         0.3   
6                          5.4                         3.6   
7                         10.4                        12.2   
8                          5.9                        16.7   
9                          9.8                         8.5   
10                        35.0                        27.2   
11                        12.8                        10.2   
12                        17.6                        10.5   
13                        11.1                         9.8   
14                        14.1                        13.1   
15                         9.1                         8.7   
16                        22.9                        25.1   
17                         5.3                         8.4   
18                        12.5                        13.5   
19                         NaN                         NaN   
20                         8.4                         8.8   
21                         2.9                         5.2   
22                         NaN                         NaN   
23                         NaN                         NaN   
24                         NaN                         NaN   
25                         NaN                         NaN   
26                         8.0                        10.3   
27                        15.3                        19.7   
28                         6.0                        11.0   
29                         NaN                         NaN   
30                        13.4                         4.2   
31                         NaN                         NaN   
32                         NaN                         NaN   
33                         NaN                         NaN   
34                        36.0                        28.7   
35                        12.0                        11.9   

    Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  
0                         12.0                         8.2  
1                          4.9                         7.7  
2                         11.2                         7.7  
3                          4.2                         5.8  
4                          NaN                         NaN  
5                          NaN                         NaN  
6                          4.4                         3.5  
7                          NaN                         NaN  
8                          6.9                        10.5  
9                          4.7                         4.9  
10                        36.5                        29.0  
11                         3.3                         9.0  
12                        11.2                        12.2  
13                         6.3                         6.4  
14                         NaN                         NaN  
15                         2.3                         4.5  
16                        26.4                        26.9  
17                         8.4                        11.1  
18                        14.4                        14.3  
19                         1.6                         3.4  
20                         3.2                         4.1  
21                         3.3                        10.1  
22                         0.2                         3.4  
23                         0.4                         1.4  
24                         0.6                         0.6  
25                         3.3                         7.1  
26                         2.3                         3.1  
2

## Statistical Analysis of dataset using pandas

In [56]:
## mean of "Scheduled Castes - 2004-05" column
data["Scheduled Castes - 2004-05"].mean()

13.180769230769231

In [57]:
## Variance of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].var()

64.55761538461539

In [58]:
## maximum value of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].max()

36.0

In [59]:
## minimum value of Scheduled Castes - 2004-05 column
data["Scheduled Castes - 2004-05"].min()

2.9

#### If u want to find Statistical analysis of numerical columns for a complete dataset

The statistics that are generated by the describe() method:

1) count tells us the number of NoN-empty rows in a feature.

2) mean tells us the mean value of that feature.

3) std tells us the Standard Deviation Value of that feature.

4) min tells us the minimum value of that feature.

5) 25%, 50%, and 75% are the percentile/quartile of each features. This quartile information helps us to detect Outliers.

6) max tells us the maximum value of that feature.

In [60]:
data.describe()

Scheduled Castes - 2004-05  Scheduled Castes - 2007-08  \
count                   26.000000                   26.000000   
mean                    13.180769                   11.700000   
std                      8.034775                    6.955171   
min                      2.900000                    0.300000   
25%                      8.575000                    8.550000   
50%                     11.550000                   10.250000   
75%                     15.000000                   13.025000   
max                     36.000000                   28.700000   

       Scheduled Tribes - 2004-05  Scheduled Tribes - 2007-08  \
count                   27.000000                   27.000000   
mean                     8.411111                    9.111111   
std                      8.793544                    6.959572   
min                      0.200000                    0.600000   
25%                      3.250000                    4.300000   
50%                      4.900000                    7.700000   
75%                     11.200000                   10.800000   
max                     36.500000                   29.000000   

       Other Backward Classes - 2004-05  Other Backward Classes - 2007-08  \
count                         27.000000                         27.000000   
mean                           9.207407                          8.481481   
std                            7.688049                          4.774696   
min                            1.500000                          3.400000   
25%                            4.500000                          5.200000   
50%                            6.400000                          7.200000   
75%                           10.150000                          9.100000   
max                           34.100000                         22.800000   

       Others - 2004-05  Others - 2007-08  All Social Groups - 2004-05  \
count         29.000000         29.000000                    36.000000   
mean           6.744828          6.855172                     8.219444   
std            4.885444          3.857792                     6.816569   
min            0.800000          0.900000                     0.400000   
25%            3.800000          4.500000                     3.700000   
50%            5.300000          5.900000                     6.300000   
75%            7.400000          9.000000                    10.425000   
max           18.400000         18.500000                    27.400000   

       All Social Groups - 2007-08  
count                    36.000000  
mean                      8.136111  
std                       5.350255  
min                       0.600000  
25%                       5.050000  
50%                       6.950000  
75%                       9.050000  
max                      25.200000

#### If u want to find Statistical description of all types of columns for a complete dataset

In [61]:
data.describe(include=['O'])

Category of States Non Special Category States
count                            36                          36
unique                            4                          36
top     Non Special Category States                   Meghalaya
freq                             19                           1